File: server/src/services/search_service.ts
import { PrismaClient } from '@prisma/client';
import { BusinessError } from '@/middlewares/error_middleware';
import { PageResult, QuoteSummary, CategorySimple } from '@/types';

const prisma = new PrismaClient();

interface OriginalTextSearchResult {
    id: string;
    title: string;
    author?: string;
    summary?: string;
    wordCount: number;
    quoteCount: number;
    matchedContent: string;
}

function extractMatchedContext(content: string, keyword: string): string {
    const lowerContent = content.toLowerCase();
    const lowerKeyword = keyword.toLowerCase();
    const idx = lowerContent.indexOf(lowerKeyword);
    if (idx === -1) {
        return content.slice(0, 80);
    }
    const start = Math.max(0, idx - 40);
    const end = Math.min(content.length, idx + keyword.length + 40);
    const before = start > 0 ? '...' : '';
    const after = end < content.length ? '...' : '';
    const snippet = content.slice(start, end);
    const escaped = keyword.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
    const highlighted = snippet.replace(
        new RegExp(escaped, 'gi'),
        (m) => `**${m}**`
    );
    return `${before}${highlighted}${after}`;
}

interface CacheEntry {
    data: string[];
    expireAt: number;
}

const hotKeywordCache = new Map<string, CacheEntry>();
const CACHE_TTL_MS = 60 * 60 * 1000;

async function searchQuotes(
    userId: string | number | null | undefined,
    keyword: string,
    page: number,
    pageSize: number
): Promise<PageResult<QuoteSummary>> {
    if (!keyword || keyword.trim().length === 0) {
        throw new BusinessError(400, '搜索关键词不能为空');
    }

    const kw = keyword.trim();
    const skip = (page - 1) * pageSize;

    const where = {
        isActive: true,
        auditStatus: 1,
        OR: [
            { content: { contains: kw } },
            { author: { contains: kw } },
            { source: { contains: kw } },
        ],
    };

    const [total, quotes] = await Promise.all([
        prisma.quote.count({ where }),
        prisma.quote.findMany({
            where,
            skip,
            take: pageSize,
            include: { quoteCategories: { include: { category: true } } },
            orderBy: { id: 'desc' },
        }),
    ]);

    const list: QuoteSummary[] = quotes.map((q) => ({
        id: String(q.id),
        content: q.content,
        author: q.author ?? undefined,
        source: q.source ?? undefined,
        summary: q.summary ?? undefined,
        isFree: q.isFree,
        categories: q.quoteCategories.map((qc) => ({
            id: String(qc.category.id),
            name: qc.category.name,
            type: qc.category.type,
        })),
    }));

    return { total, list };
}

async function getHotKeywords(): Promise<string[]> {
    const cached = hotKeywordCache.get('hot');
    if (cached && cached.expireAt > Date.now()) {
        return cached.data;
    }

    const results = await prisma.$queryRaw<Array<{ keyword: string; count: number }>>`
        SELECT keyword, COUNT(*) as count
        FROM SearchHistory
        GROUP BY keyword
        ORDER BY count DESC
        LIMIT 10
    `;

    const keywords = results.map((r) => r.keyword);
    hotKeywordCache.set('hot', { data: keywords, expireAt: Date.now() + CACHE_TTL_MS });

    return keywords;
}

async function searchOriginalText(
    keyword: string,
    page: number,
    pageSize: number
): Promise<PageResult<OriginalTextSearchResult>> {
    if (!keyword || keyword.trim().length === 0) {
        throw new BusinessError(400, '搜索关键词不能为空');
    }

    const kw = keyword.trim();
    const skip = (page - 1) * pageSize;

    const where = {
        auditStatus: 1,
        OR: [
            { title: { contains: kw } },
            { content: { contains: kw } },
            { author: { contains: kw } },
        ],
    };

    const [total, records] = await Promise.all([
        prisma.originalText.count({ where }),
        prisma.originalText.findMany({
            where,
            skip,
            take: pageSize,
            orderBy: { id: 'desc' },
        }),
    ]);

    const list: OriginalTextSearchResult[] = records.map((r) => ({
        id: String(r.id),
        title: r.title,
        author: r.author ?? undefined,
        summary: r.summary ?? undefined,
        wordCount: r.wordCount,
        quoteCount: r.quoteCount,
        matchedContent: extractMatchedContext(r.content, kw),
    }));

    return { total, list };
}

async function saveHistory(userId: number, keyword: string): Promise<void> {
    if (!keyword || keyword.trim().length === 0) return;

    const kw = keyword.trim();

    await prisma.searchHistory.deleteMany({
        where: { userId, keyword: kw },
    });

    await prisma.searchHistory.create({
        data: { userId, keyword: kw },
    });
}

async function clearHistory(userId: number): Promise<void> {
    await prisma.searchHistory.deleteMany({ where: { userId } });
}

const getHot = getHotKeywords;
export { searchQuotes, searchOriginalText, getHotKeywords, getHot, saveHistory, clearHistory };

File: server/app/services/search_service.py
"""搜索服务

对应 TypeScript 版 server/src/services/search_service.ts
"""
import re
import time
from datetime import datetime

from sqlalchemy import or_, func
from sqlalchemy.orm import Session

from ..deps import BusinessError
from ..models import Quote, OriginalText, SearchHistory, QuoteCategory, Category
from ..utils.logger import logger

_HOT_KEYWORD_CACHE: dict = {}
_CACHE_TTL_MS = 60 * 60 * 1000


def _extract_matched_context(content: str, keyword: str) -> str:
    lower_content = content.lower()
    lower_keyword = keyword.lower()
    idx = lower_content.find(lower_keyword)
    if idx == -1:
        return content[:80]
    start = max(0, idx - 40)
    end = min(len(content), idx + len(keyword) + 40)
    before = "..." if start > 0 else ""
    after = "..." if end < len(content) else ""
    snippet = content[start:end]
    escaped = re.escape(keyword)
    highlighted = re.sub(escaped, lambda m: f"**{m.group()}**", snippet, flags=re.IGNORECASE)
    return f"{before}{highlighted}{after}"


def _get_categories_for_quote(db: Session, quote_id: int) -> list:
    qcs = db.query(QuoteCategory).filter(QuoteCategory.quoteId == quote_id).all()
    if not qcs:
        return []
    cat_ids = [qc.categoryId for qc in qcs]
    cats = db.query(Category).filter(Category.id.in_(cat_ids)).all()
    return [{"id": str(c.id), "name": c.name, "type": c.type} for c in cats]


def searchQuotes(
    db: Session,
    user_id: int | None = None,
    keyword: str = "",
    page: int = 1,
    pageSize: int = 20,
) -> dict:
    if not keyword or not keyword.strip():
        raise BusinessError(400, "搜索关键词不能为空")

    kw = keyword.strip()
    skip = (page - 1) * pageSize

    conditions = [
        Quote.isActive == True,
        Quote.auditStatus == 1,
        or_(
            Quote.content.contains(kw),
            Quote.author.contains(kw),
            Quote.source.contains(kw),
        ),
    ]

    total = db.query(func.count(Quote.id)).filter(*conditions).scalar()
    quotes = (
        db.query(Quote)
        .filter(*conditions)
        .order_by(Quote.id.desc())
        .offset(skip)
        .limit(pageSize)
        .all()
    )

    # Batch load categories for all quotes
    quote_ids = [q.id for q in quotes]
    cat_map: dict[int, list] = {}
    if quote_ids:
        qcs = db.query(QuoteCategory).filter(QuoteCategory.quoteId.in_(quote_ids)).all()
        cat_ids = list({qc.categoryId for qc in qcs})
        cats = {c.id: c for c in db.query(Category).filter(Category.id.in_(cat_ids)).all()} if cat_ids else {}
        for qc in qcs:
            c = cats.get(qc.categoryId)
            if c:
                cat_map.setdefault(qc.quoteId, []).append(
                    {"id": str(c.id), "name": c.name, "type": c.type}
                )

    list_data = [
        {
            "id": str(q.id),
            "content": q.content,
            "author": q.author or None,
            "source": q.source or None,
            "summary": q.summary or None,
            "isFree": q.isFree,
            "categories": cat_map.get(q.id, []),
        }
        for q in quotes
    ]

    return {"total": total, "list": list_data}


def getHotKeywords(db: Session) -> list:
    cached = _HOT_KEYWORD_CACHE.get("hot")
    if cached and cached["expireAt"] > int(time.time() * 1000):
        return cached["data"]

    results = (
        db.query(SearchHistory.keyword, func.count(SearchHistory.keyword).label("cnt"))
        .group_by(SearchHistory.keyword)
        .order_by(func.count(SearchHistory.keyword).desc())
        .limit(10)
        .all()
    )

    keywords = [r[0] for r in results]
    _HOT_KEYWORD_CACHE["hot"] = {"data": keywords, "expireAt": int(time.time() * 1000) + _CACHE_TTL_MS}

    return keywords


getHot = getHotKeywords


def searchOriginalText(
    db: Session,
    keyword: str,
    page: int = 1,
    pageSize: int = 20,
) -> dict:
    if not keyword or not keyword.strip():
        raise BusinessError(400, "搜索关键词不能为空")

    kw = keyword.strip()
    skip = (page - 1) * pageSize

    conditions = [
        OriginalText.auditStatus == 1,
        or_(
            OriginalText.title.contains(kw),
            OriginalText.content.contains(kw),
            OriginalText.author.contains(kw),
        ),
    ]

    total = db.query(func.count(OriginalText.id)).filter(*conditions).scalar()
    records = (
        db.query(OriginalText)
        .filter(*conditions)
        .order_by(OriginalText.id.desc())
        .offset(skip)
        .limit(pageSize)
        .all()
    )

    list_data = [
        {
            "id": str(r.id),
            "title": r.title,
            "author": r.author or None,
            "summary": r.summary or None,
            "wordCount": r.wordCount,
            "quoteCount": r.quoteCount,
            "matchedContent": _extract_matched_context(r.content, kw),
        }
        for r in records
    ]

    return {"total": total, "list": list_data}


def saveHistory(db: Session, userId: int, keyword: str) -> None:
    if not keyword or not keyword.strip():
        return

    kw = keyword.strip()

    db.query(SearchHistory).filter(
        SearchHistory.userId == userId,
        SearchHistory.keyword == kw,
    ).delete()
    db.commit()

    record = SearchHistory(userId=userId, keyword=kw)
    db.add(record)
    db.commit()


def clearHistory(db: Session, userId: int) -> None:
    db.query(SearchHistory).filter(SearchHistory.userId == userId).delete()
    db.commit()
